apis/projectcontour/v1/httpproxy.go

```diff
@@ -146,6 +149,34 @@ type TLS struct {
 	EnableFallbackCertificate bool `json:"enableFallbackCertificate,omitempty"`
 }
 
+// CORSHeaderValue specifies the value of the string headers returned by a cross-domain request.
+// +kubebuilder:validation:Pattern=^[a-zA-Z0-9!#$%&'*+.^_\x60|~-]+$
+type CORSHeaderValue string
+
+// CORSPolicy allows setting the CORS policy
+type CORSPolicy struct {
+	// Specifies whether the resource allows credentials.
+	//  +optional
+	AllowCredentials bool `json:"allowCredentials,omitempty"`
+	// AllowOrigin specifies the origins that will be allowed to do CORS requests.
+	AllowOrigin []string `json:"allowOrigin"`
+	// AllowMethods specifies the content for the *access-control-allow-methods* header.
+	AllowMethods []CORSHeaderValue `json:"allowMethods"`
+	// AllowHeaders specifies the content for the *access-control-allow-headers* header.
+	//  +optional
+	AllowHeaders []CORSHeaderValue `json:"allowHeaders,omitempty"`
+	// ExposeHeaders Specifies the content for the *access-control-expose-headers* header.
+	//  +optional
+	ExposeHeaders []CORSHeaderValue `json:"exposeHeaders,omitempty"`
+	// MaxAge indicates for how long the results of a preflight request can be cached.
+	// MaxAge durations are expressed in the Go [Duration format](https://godoc.org/time#ParseDuration).
+	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+	// Only possitive values are allowed while 0 disables the cache requiring a preflight OPTIONS
+	// check for all cross-origin requests.
+	//  +optional
+	MaxAge string `json:"maxAge,omitempty"`
```

Comment:
Normally, users have 3 different options for fields like this:

1. Specify a positive duration - Contour configures that specific value in Envoy
2. Don't specify a value (or specify 0) - Contour **does not** configure any value for that field in Envoy, which tells Envoy to use its default behavior
3. Specify `infinity` - Contour (usually) explicitly passes a 0 to Envoy for that field, which tells Envoy to disable the timeout in question.

It'd be best if we could have similar semantics for this field. The challenge here seems to be that the word "infinity" doesn't really make sense to mean "disable caching".

I can think of a few different options:
- empty string means "don't configure the field", explicit "0" means "disable caching"
- empty string or 0 means "don't configure the field", "-1" or any negative duration string means "disable caching"
- empty string or 0 means "don't configure the field", the string "disabled" means "disable caching"

Personally I think I prefer the third of these (i.e. support "disabled" as a magic string), but I'd like to get @stevesloka  and @youngnick's opinions here.

Reply:
I prefer the first option because in most cache systems 0 means "disable caching" and using that value for "don't configure the field" could be misleading.

Reply:
That's a fair point @aberasarte, but I'd also like to keep it a bit consistent with the rest of Contour. WDYT about "0" or "disabled" means "disable caching"? If that's a no, I'm okay with just "0".

Reply:
> WDYT about "0" or "disabled" means "disable caching"?

That sounds good to me but then I would need a way to identify the "don't configure the field" option. `UseDefault`  would work [but at the moment 0 is treated as UseDefault](https://github.com/projectcontour/contour/blob/main/internal/timeout/timeout.go#L37) and that's not what we want here. 

I could create a different package (`maxage`? `expiration`?) for handling this kind of values. It would be similar to https://github.com/projectcontour/contour/blob/main/internal/timeout/timeout.go but it would allow the "disabled" value instead of "infinity" and it wouldn't treat `0` as `UseDefault`. WDYT?

Reply:
This seems pretty clear to me. The field is called `MaxAge`, so a maximum age of 0 would disable caching, and a maximum age of infinity would cache forever. To distinguish between the default and 0 as an explictly specified age, you can use `MaxAge *string`. It also seems pretty clear that a value of "" would have a different result than a value of "0s".

Reply:
> The field is called MaxAge, so a maximum age of 0 would disable caching, and a maximum age of infinity would cache forever.

I wouldn't allow `infinity` as it's potentially dangerous. Browsers even have an imposed limit for this value. [From the spec](https://fetch.spec.whatwg.org/#cors-preflight-fetch):

> If max-age is greater than an imposed limit on max-age, then set max-age to the imposed limit.

I think that we should encourage people to use lower values.

> To distinguish between the default and 0 as an explictly specified age, you can use MaxAge *string. It also seems pretty clear that a value of "" would have a different result than a value of "0s".

I totally agree. The problem is that we turn that string into a Golang duration, so we need a way for differentiating whether that value has been set. At the moment https://github.com/projectcontour/contour/blob/main/internal/timeout/timeout.go is used for the same purpose but for other properties. I was trying to use the same logic for `MaxAge` but if we want to allow the "disabled" value without breaking the current behavior, it would probably be better to create a different utility package.

Reply:
I'm satisfied with "0s" explicitly specified means "disabled", "" (no value specified) means "use default". We'll need custom parse logic instead of the standard `timeout.Parse` to properly handle those cases, but we can still use `timeout.Setting` values internally to represent it.

Reply:
Yes, that seems fine to me.

Reply:
>I'm satisfied with "0" explicitly specified means "disabled", "" (no value specified) means "use default". We'll need custom parse logic instead of the standard timeout.Parse to properly handle those cases, but we can still use timeout.Setting values internally to represent it. I don't think we need to use a string pointer either, we can just check whether the field was (a) empty or (b) a non-empty representation of "0".

Perfect, then we need to change the logic [of the UseDefault() method](https://github.com/projectcontour/contour/blob/main/internal/timeout/timeout.go#L37) as well. Instead of checking `s.val == 0` for deciding whether a value has been set, I could add a new property to the `Setting` struct for that. Something in the lines of:

```
type Setting struct {
	val        time.Duration
	disabled   bool
	usedefault bool
}
```

And check that property in the `UseDefault` method:

```
func (s Setting) UseDefault() bool {
	return !s.disabled && s.usedefault
}
```

The current `Parse` function would set that property to `true` if the value is `0`, so nothing should break. I would add a new parsing function (`ParseMaxeAge`?) that would be similar to `Parse` but this one would set `usedefault` to `true` only if the received string is empty `""`.

Would that work for you?

Reply:
Yep, that sounds reasonable to me.

Reply:
I inverted the logic in order to make the change backwards compatible. The new property is called `valset`.